### ISSUE TITLE
[question] Is caffeine virtual thread / loom friendly?

### ISSUE THREAD

**Issue by wyhasany** [ben-manes/caffeine#1468]
I wonder if caffeine is loom-friendly - e.g. does not pin the platform threads?

**Comment by He-Pin** [comment]
Yes and no
Concurrenthashmap is blocking the carrier thread. They plan to fix it,but not jdk23

**Comment by ben-manes** [comment]
Currently in Java, `synchronized` pins the virtual thread. That is being worked on currently and will be resolved soonish ([status](https://www.youtube.com/watch?v=WoQJnnMIlFY&t=260s)).

Caffeine uses `ConcurrentHashMap` which is built on `synchronized` blocks, such as for `computeIfAbsent`. Since the issue will be resolved soon and the Java team is not rewriting all of their collections towards `ReentrantLock`, we'll following their lead by waiting for the platform to catch up. There are a few other areas where virtual threads are penalized compared to platform threads such as the lack of preemption, network address resolution ([JEP 418](https://openjdk.org/jeps/418)), need for better async io (io_ring), scalable timers, and more. It will be a long transition until VT are preferred in all cases.

In the meantime, you can use AsyncCache which decouples the computation from the map operation. Then the map's synchronized blocks are short cpu work and the I/O is performed by a `CompletableFuture` (which is VT friendly). See 
https://github.com/ben-manes/caffeine/issues/779 for an example.

**Comment by He-Pin** [comment]
Or maybe just set a larger pool size。

**Comment by wyhasany** [comment]
@ben-manes `AsyncCache` seems to be a perfect solution from my perspective 🎉  do you plan to rewrite `synchronized` blocks to `ReentrantLocks` in caffeine codebase?

**Comment by He-Pin** [comment]
I think we may better follow Hotspot team's lead .

**Comment by ben-manes** [comment]
Since `ConcurrentHashMap` will pin the thread in its synchronized blocks, there is not much of an incentive to change Caffeine's usage. Theirs is per-hashbin whereas ours is per-entry. For short cpu work needed by AsyncCache it is perfectly fine to pin as VTs do not preempt. For a synchronized cache then ConcurrentHashMap's compute will pin regardless, so any effort we make will not be helpful.

A few years ago I did chat with Doug Lea who was pondering what a change to ConcurrentHashMap might look like. He was mused _separate bitwise-trie-like structure of AQS objects that serve as the inflated forms. One TBD is how/when to kill them off_, where AQS is the synchronization primitive underlying ReentrantLock. I haven't seen any progress there the Loom github shows steady progress on mutex support.

You might follow [loom-dev](https://mail.openjdk.org/pipermail/loom-dev/) and the [github branches](https://github.com/openjdk/loom/branches) to keep an eye on progress.

**Comment by He-Pin** [comment]
I just upgrade my system to jdk21 and use both virtual thread, and removed some old guava code and switch to caffeine, after I trace the pin.


I was try to submit pr to cchm too , but seems not much good way, otherwise it will wast memory. 

I also asked on Loom dev, they are working on it , but no timeline，and cchm will not be rewritten. 

I see 10％ cpu reduce and 5% mem reduce on 16core 32gb memory box, and at the workload 60% and where with VT is 50% on the same workload. This is a mission critical system of taobao live.

That's what I can share and I think you can just follow author's advice, that what we can best do for now.

**Comment by ben-manes** [comment]
FYI - The new EA builds that no longer pin on synchronization ([announcement](https://mail.openjdk.org/pipermail/loom-dev/2024-February/006433.html)) and they are asking for usage feedback.

**Comment by jkvargas** [comment]
hi, quick question
what is the actual status of caffeine and virtual threads?

**Comment by ben-manes** [comment]
Per the link above, the Java team's work should resolve this in an upcoming release. You can try their early access build to provide feedback. For the time being, virtual threads should be used judiciously in limited scenarios until many of the show stopper problems have been resolved by the JDK. Likely the next LTS (25) will be a good time to consider using it.

**Comment by anklos** [comment]
hey @ben-manes we also run into the same problem with JDK21. I wonder could we replace `Spymemcached` with `XMemcached` to fix the issue? Please see the attached pined thread stack trace

<img width="1154" height="262" alt="Image" src="https://github.com/user-attachments/assets/ce90fa54-a666-43cb-af65-6a44fafa6f99" />

**Comment by ben-manes** [comment]
I'm not sure how that relates to Caffeine, but I don't think using Virtual Threads in JDK-21 is safe and in JDK-25, while safe, its still unwise. It is fast but pins easily and is non-preemptive, which can cause a lot of idiosyncratic behavior.

Since its an async call anyway, you could replace`getBulk` with `asyncGetBulk` and use `AsyncCache`. I thought Spymemcached was abandoned a decade ago or so, so XMemcached might make sense just because its maintained.

**Comment by anklos** [comment]
sorry just realise we do ` Our CacheLoader -> Caffeine L1 or Spymemcached L2`, you are right it is not really about Caffeine. 

yeah in this case we might able to easily replace that.

**Comment by anklos** [comment]
just found our Caffeine also has the same issue in other pined threads stack traces.

>  I don't think using Virtual Threads in JDK-21 is safe and in JDK-25, while safe, its still unwise. 

because our Tomcat server is configured to use virtual thread, our jvm app also uses that to avoid the overhead of double scheduling onto platform threads.

do you think we should switch everything to platform threads?

**Comment by ben-manes** [comment]
JDK-21 pins threads on synchronized blocks, which are everywhere in the standard library. Its not feasible to avoid pinning on them for the Tomcat server threads, e.g. due to `ConcurrentHashMap`. It is acceptable for very short cases without contention, so AsyncCache can work since it performs the expensive load within a `Future` which does not itself pin. But the reality is nothing serious will work in an application with Virtual Threads until JDK-24 when that issue was fixed. If you can move to JDK-25 (latest LTS) then it is worth a try.

In my experience on JDK-25 it still is a poor fit for a web server. If your server uses brotli or zstd compression then that will pin the thread in native code and cause large latency spikes. Likewise you'll see an increase in worst-case latency because VTs cannot be pre-empted, so a compute intensive task can seriously degrade user-facing ones. It might make sense in a narrowly focused micro-service.

I think we are still a few years away before its safe to turn on generally. It is okay for pockets where you know the flows are compatible but modestly sized applications have too many unknowns and code evolution will change it that you'll run into those flaws and suffer bad performance.

**Comment by mattrpav** [comment]
@ben-manes curiously, the VT executor service will spin up add'l threads. Are you seeing the high-latency once the max os threads is reached (default: 256)?

```
jdk.virtualThreadScheduler.parallelism: Sets the number of platform threads (carrier threads) available to the scheduler. This defaults to the number of available processors.

jdk.virtualThreadScheduler.maxPoolSize: Sets the maximum number of platform threads the scheduler can create. The default is typically 256.

jdk.virtualThreadScheduler.minRunnable:S pecifies the minimum number of core threads to keep alive in the scheduler's pool.
```

**Comment by ben-manes** [comment]
@mattrpav I saw high latencies in a production-facing monolithic server when switching to virtual threads. iirc, Jetty uses a default threadpool size of 200 and a larger number of threads were in brotli native code for the responses. This server has many other thread pools, e.g. pub-sub work queues, so its likely that the virtualThreadScheduler defaults needed to be adjusted. It was an easy experiment swap and it appeared to increase latencies so we reverted, but I don't recall much anymore. There were various other performance work done before or after that could explain it, such as discovering GCP's load balancer defaulted to being geo-IP based to create unbalanced load (overload one server to then fall back to the next, creating domino effects and latency spikes; reconfigured to  be round robin). A monolithic server might not have been a good experiment since the workloads commingle, whereas more focused microservices tend to be easier to debug and tune.
